File: js/link_queue/link_queue.js
function QNode(props){
    this.data = props.data
    this.next = props.next
}

function LinkQueue(props){
    this.front = new QNode({})
    this.rear = new QNode({})
}

LinkQueue.prototype.visit=function(element){
    process.stdout.write(element + ' ')
}

// Construct an empty queue
LinkQueue.prototype.InitQueue = function(){
    this.front = this.rear = new QNode({})
	if (!this.front)
		return false
    this.front.next = null
    return true
}

// Destroy queue
LinkQueue.prototype.DestroyQueue = function()
{
	while (this.front)
	{
		this.rear = this.front.next
		this.front = this.rear
	}
	return true
}

// Clear queue
LinkQueue.prototype.ClearQueue = function()
{
	var p, q
	this.rear = this.front
	p = this.front.next
	this.front.next = null
	while (p)
	{
		q = p
		p = p.next
		q = null
	}
	return true
}

// If it is empty queue,return TRUE,otherwise return FALSE
LinkQueue.prototype.QueueEmpty = function()
{
	if (this.front == this.rear)
		return true
	else
		return false
}

// Get the length of queue
LinkQueue.prototype.QueueLength = function()
{
	var i = 0
	var p
	p = this.front
	while (this.rear != p)
	{
		i++
		p = p.next
	}
	return i
}

// If queue is not empty,return head element of queue and return OK,otherwise return ERROR
LinkQueue.prototype.GetHead = function()
{
	var p
	if (this.front == this.rear)
		return false
	p = this.front.next
	e = p.data
    return e
}

// Insert element as new rear element
LinkQueue.prototype.EnQueue = function(e)
{
	var s = new QNode({})
	if (!s)
		return false
	s.data = e
	s.next = null
	// Assign new node to next of former rear node
	this.rear.next = s
	// Assign s to rear node
	this.rear = s
	return true
}

// If queue is not empty,delete head element of queue,then return deleted value,otherwise return false
LinkQueue.prototype.DeQueue = function(e)
{
	var p
	if (this.front == this.rear)
		return false
	p = this.front.next
	e = p.data
	this.front.next = p.next
	// Judge if the first node is rear
	if (this.rear == p)
        this.rear = this.front
	return e
}

// Traverse every element from front to rear in turn
LinkQueue.prototype.QueueTraverse = function()
{
	var p
	p = this.front.next
	while (p)
	{
		this.visit(p.data)
		p = p.next
	}
    console.log()
	return true
}

var lq = new LinkQueue({})
var i = lq.InitQueue()
if (i)
    console.log("Construct an empty queue!\n")
console.log("If queue is empty? \n", lq.QueueEmpty())
console.log("Queue length is %d\n", lq.QueueLength())
lq.EnQueue(-5)
lq.EnQueue(5)
lq.EnQueue(10)
console.log("After inserting 3 element(-5,5,10),Queue length is %d\n" ,lq.QueueLength())
console.log("If queue is empty?\n", lq.QueueEmpty())
console.log("Elements of queue are in order:")
lq.QueueTraverse()
d = lq.GetHead()
if (i == true)
    console.log("Head element of queue is:%d\n", d)
dq = lq.DeQueue()
console.log("After deleting head element of queue,deleted element is %d\n", dq)
i = lq.GetHead()
if (i == true)
    console.log("New head element of queue is:%d\n", d)
lq.ClearQueue()
console.log("After clearing queue,q.front=%d q.rear=%d q.front->next=%d\n", lq.front, lq.rear, lq.front.next)
console.log("Queue length is %d\n", lq.QueueLength())
lq.DestroyQueue()
console.log("After destroying,q.front=%d q.rear=%d\n", lq.front, lq.rear)
console.log("Queue length is %d\n", lq.QueueLength())

File: c/link_queue/link_queue/link_queue.cpp
#include "stdio.h"    
#include "stdlib.h"   
#include "io.h"  
#include "math.h"  
#include "time.h"

#define OK 1
#define ERROR 0
#define TRUE 1
#define FALSE 0
// Define capacity
#define MAXSIZE 20

typedef int Status;
// Define data type
typedef int QElemType;

typedef struct QNode
{
	QElemType data;
	struct QNode *next;
}QNode, *QueuePtr;

typedef struct
{
	QueuePtr front, rear; 
}LinkQueue;

Status visit(QElemType c)
{
	printf("%d ", c);
	return OK;
}

// Construct an empty queue
Status InitQueue(LinkQueue *Q)
{
	Q->front = Q->rear = (QueuePtr)malloc(sizeof(QNode));
	if (!Q->front)
		exit(OVERFLOW);
	Q->front->next = NULL;
	return OK;
}

// Destroy queue
Status DestroyQueue(LinkQueue *Q)
{
	while (Q->front)
	{
		Q->rear = Q->front->next;
		free(Q->front);
		Q->front = Q->rear;
	}
	return OK;
}

// Clear queue
Status ClearQueue(LinkQueue *Q)
{
	QueuePtr p, q;
	Q->rear = Q->front;
	p = Q->front->next;
	Q->front->next = NULL;
	while (p)
	{
		q = p;
		p = p->next;
		free(q);
	}
	return OK;
}

// If Q is empty queue,return TRUE,otherwise return FALSE
Status QueueEmpty(LinkQueue Q)
{
	if (Q.front == Q.rear)
		return TRUE;
	else
		return FALSE;
}

// Get the length of queue
int QueueLength(LinkQueue Q)
{
	int i = 0;
	QueuePtr p;
	p = Q.front;
	while (Q.rear != p)
	{
		i++;
		p = p->next;
	}
	return i;
}

// If queue is not empty,return head element of queue and return OK,otherwise return ERROR
Status GetHead(LinkQueue Q, QElemType *e)
{
	QueuePtr p;
	if (Q.front == Q.rear)
		return ERROR;
	p = Q.front->next;
	*e = p->data;
	return OK;
}

// Insert element as new rear element
Status EnQueue(LinkQueue *Q, QElemType e)
{
	QueuePtr s = (QueuePtr)malloc(sizeof(QNode));
	if (!s)
		exit(OVERFLOW);
	s->data = e;
	s->next = NULL;
	// Assign new node to next of former rear node
	Q->rear->next = s;
	// Assign s to rear node
	Q->rear = s;
	return OK;
}

// If queue is not empty,delete head element of queue,then return deleted value and return OK,otherwise return ERROR
Status DeQueue(LinkQueue *Q, QElemType *e)
{
	QueuePtr p;
	if (Q->front == Q->rear)
		return ERROR;
	p = Q->front->next;		
	*e = p->data;
	Q->front->next = p->next;
	// Judge if the first node is rear
	if (Q->rear == p)
		Q->rear = Q->front;
	free(p);
	return OK;
}

// Traverse every element from front to rear in turn
Status QueueTraverse(LinkQueue Q)
{
	QueuePtr p;
	p = Q.front->next;
	while (p)
	{
		visit(p->data);
		p = p->next;
	}
	printf("\n");
	return OK;
}

int main()
{
	int i;
	QElemType d;
	LinkQueue q;
	i = InitQueue(&q);
	if (i)
		printf("Construct an empty queue!\n");
	printf("If queue is empty? %d(1:Empty 0:Not Empty)\n", QueueEmpty(q));
	printf("Queue length is %d\n", QueueLength(q));
	EnQueue(&q, -5);
	EnQueue(&q, 5);
	EnQueue(&q, 10);
	printf("After inserting 3 element(-5,5,10),Queue length is %d\n", QueueLength(q));
	printf("If queue is empty?%d(1:Empty 0:Not Empty)\n", QueueEmpty(q));
	printf("Elements of queue are in order:");
	QueueTraverse(q);
	i = GetHead(q, &d);
	if (i == OK)
		printf("Head element of queue is:%d\n", d);
	DeQueue(&q, &d);
	printf("After deleting head element of queue%d\n", d);
	i = GetHead(q, &d);
	if (i == OK)
		printf("New head element of queue is:%d\n", d);
	ClearQueue(&q);
	printf("After clearing queue,q.front=%u q.rear=%u q.front->next=%u\n", q.front, q.rear, q.front->next);
	printf("Queue length is %d\n", QueueLength(q));
	DestroyQueue(&q);
	printf("After destroying,q.front=%u q.rear=%u\n", q.front, q.rear);
	printf("Queue length is %d\n", QueueLength(q));
	system("pause");
	return 0;
}